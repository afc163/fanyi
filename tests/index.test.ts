import { fork } from 'node:child_process';
import path from 'node:path';
import { describe, expect, it } from 'vitest';

const scriptPath = path.resolve(__dirname, '../bin/fanyi.js');

const runScript = (args: string[] = []): Promise<{ stdout: string; stderr: string }> => {
  return new Promise((resolve, reject) => {
    const child = fork(scriptPath, args, { silent: true });

    let stdout = '';
    let stderr = '';

    child.stdout?.on('data', (data) => {
      stdout += data.toString();
    });

    child.stderr?.on('data', (data) => {
      stderr += data.toString();
    });

    child.on('close', (code) => {
      if (code !== 0) {
        reject({ code, stdout, stderr });
      } else {
        resolve({ stdout, stderr });
      }
    });
  });
};

describe('fanyi CLI', () => {
  it('should print translation of the word', async () => {
    const { stdout } = await runScript(['hello']);
    expect(stdout).toContain(`hello  英[ hə'ləʊ ]  美[ həˈloʊ ]  ~  iciba.com`);
  });

  it('should print usage if no arguments are given', async () => {
    const { stdout } = await runScript();
    expect(stdout).toMatchSnapshot();
  });

  it('should print help if -h is given', async () => {
    const { stdout } = await runScript(['-h']);
    expect(stdout).toMatchSnapshot();
  });

  it('should be able to config global options', async () => {
    const { stdout } = await runScript(['config', 'set', 'color', 'false']);
    expect(stdout).toContain('{"color":false}');
    const { stdout: stdout2 } = await runScript(['config', 'set', 'color', 'true']);
    expect(stdout2).toContain('{"color":true}');
    const { stdout: stdout3 } = await runScript(['config', 'set', 'iciba', 'false']);
    expect(stdout3).toContain('{"iciba":false}');
    const { stdout: stdout4 } = await runScript(['config', 'set', 'iciba', 'true']);
    expect(stdout4).toContain('{"iciba":true}');
  });

  it('should print config', async () => {
    const { stdout } = await runScript(['config', 'list']);
    expect(stdout).toContain('.config/fanyi/.fanyirc');
  });

  it('should print search history', async () => {
    const { stdout } = await runScript(['list']);
    expect(stdout).toContain('fanyi history:');
  });
});
