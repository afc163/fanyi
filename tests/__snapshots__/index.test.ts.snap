// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fanyi CLI > should print help if -h is given 1`] = `
"Usage: fanyi [options] [command]

A translate tool in your command line

Options:
  -V, --version   output the version number
  -h, --help      display help for command

Commands:
  config          设置全局选项
  list [options]

示例:
  $ fanyi word
  $ fanyi world peace
  $ fanyi chinglish
  $ fanyi config set color true
  $ fanyi config set iciba true
  $ fanyi config set groq true
  $ fanyi config set GROQ_API_KEY your_api_key_here
  $ fanyi config list

"
`;

exports[`fanyi CLI > should print usage if no arguments are given 1`] = `
"Usage: fanyi [options] [command]

A translate tool in your command line

Options:
  -V, --version   output the version number
  -h, --help      display help for command

Commands:
  config          设置全局选项
  list [options]

示例:
  $ fanyi word
  $ fanyi world peace
  $ fanyi chinglish
  $ fanyi config set color true
  $ fanyi config set iciba true
  $ fanyi config set groq true
  $ fanyi config set GROQ_API_KEY your_api_key_here
  $ fanyi config list

"
`;
